### PR TITLE
Implement RFE 5361: enhance Princess handling of hazardous terrain

### PR DESCRIPTION
**Pull request by Sleet01**
Adds additional hazard analysis for:
- Magma (Crust and Liquid, aka "Lava")
- Swamp
- Mud
- Water (for Submarine / UMU units)

in order to prevent Princess from exposing some unit types to excessive hazard while choosing paths.
Implements RFE #5361 

Testing:
- Ran some Bot v Bot scenarios with the relevant Terrains and movement types.
- Added new unit tests for specific interactions.
- Ran all 3 projects' unit tests.

Note: UMU and Submarine Infantry units can now traverse Water hexes more easily, but pathing in/out of Water hexes seems to be broken.

### CODE REVIEW

**Comment by Thom293** [comment]
Thank you Sleet.  Will test in a few hours. 

I believe this will also close #4902 ?

**Comment by HammerGS** [comment]
Does this mean that Princess would be able to operate on water fully now?

**Comment by Thom293** [comment]
RE: Pathing in and out of water.  I dont know for sure.  Maybe it was broken already.  But I did make some heavy handed changes before to Ice and PSR tolerance that may have some influence.  Maybe not, but for full disclosure, here they are if someone wants to see if I broke the water pathing:

https://github.com/MegaMek/megamek/pull/5041
https://github.com/MegaMek/megamek/pull/5111

**Comment by Sleet01** [comment]
> Does this mean that Princess would be able to operate on water fully now?

I definitely saw UMU, Orca, and SCUBA units moving in and out of water, and at their full MP within water.
The water pathing may still need work, though: I only saw a couple of units exit to land (or enter the water _from_ land) in my testing.

**Comment by Sleet01** [comment]
> Thank you Sleet. Will test in a few hours.
> 
> I believe this will also close #4902 ?

I believe so; I'll keep testing when I can.

**Comment by repligator** [comment]
Everything is a definite improvement so far. Princess still makes some questionable decisions (for example choosing to walk over the hex where the magma crust has already cracked instead of one where it hasn't), but the important part (seeing the terrain as hazardous in the first place) is working.

**Comment by Sleet01** [comment]
> Everything is a definite improvement so far. Princess still makes some questionable decisions (for example choosing to walk over the hex where the magma crust has already cracked instead of one where it hasn't), but the important part (seeing the terrain as hazardous in the first place) is working.

I think I can explain this as: there's a chance to get stuck if walking over a Crust hex, if the crust hex cracks.  That's a 1/6 chance, but then the expected damage from getting stuck is (Damage from Lava * (1.0 + chance of getting stuck + number of turns to escape)).  So it's very possible for the heat and damage from walking through Liquid Magma to be a _lower_ risk than possibly getting stuck after the Magma Crust breaks, especially for lower-PS pilots.

**Comment by Sleet01** [comment]
Added increasing hazard from Lava as unit takes damage.

**Comment by Sleet01** [comment]
It looks like no amount of damage, heat, or hazard will prevent determined Berserker Princess units from at least dipping their toes in some Liquid Magma, especially now that we've removed the (erroneously applied) +4 PSR to bog down when just crossing Liquid Magma hexes.

But the latest update should dissuade all but the most bloodthirsty.

File: megamek/src/megamek/client/bot/princess/BasicPathRanker.java [hunk 1]
```diff
@@ -959,6 +1004,12 @@ private double calcWaterHazard(Entity movingUnit, Hex hex, MoveStep step, MovePa
             return 0;
         }
 
+        // Submarine units should be fine; Orca-riding Infantry goes here.
+        if (EntityMovementMode.SUBMARINE == movingUnit.getMovementMode()) {
```
Comment (SJuliez) [inline megamek/src/megamek/client/bot/princess/BasicPathRanker.java:1008]: the movement mode has some useful bool methods, .isSubmarine()

File: megamek/src/megamek/common/Terrain.java [hunk 2]
```diff
@@ -607,6 +607,7 @@ public int getBogDownModifier(EntityMovementMode moveMode, boolean largeVee) {
                     return 0;
                 }
             case MAGMA:
+                // Only applies when jumping into a hex.
```
Comment (SJuliez) [inline megamek/src/megamek/common/Terrain.java:610]: seeing this and that the method already takes some additional parameters it's questionable if this should actually be in Terrain at all. Nothing's easy in BT. (Just commenting, no action required)

**Review by SJuliez — APPROVED** [review]
looks good to me. EntityMovementMode has some nice boolean methods (e.g. isHoverOrWiGE()) but those can be refactored in at another point.

**Pull request merged**